src/bson.c

```diff
@@ -617,10 +621,22 @@ void object_to_bson(zval *object, php_phongo_bson_flags_t flags, const char *key
 				return;
 			}
 
-			if (Z_TYPE_P(obj_data) != IS_ARRAY) {
-				phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE TSRMLS_CC, "Expected %s() to return an array, %s given", BSON_SERIALIZE_FUNC_NAME, zend_get_type_by_const(Z_TYPE_P(obj_data)));
+			if (Z_TYPE_P(obj_data) != IS_ARRAY && !(Z_TYPE_P(obj_data) == IS_OBJECT && instanceof_function(Z_OBJCE_P(obj_data), zend_standard_class_def TSRMLS_CC))) {
+				const char *classname = NULL;
+				zend_uint classname_len = 0;
+				int classname_free = 0;
+
+				if (Z_TYPE_P(obj_data) == IS_OBJECT) {
+					classname_free = !zend_get_object_classname(obj_data, &classname, &classname_len TSRMLS_CC);
```

Comment:
@bjori: I noticed you were `Z_OBJCE_P(object)->name` directly when creating the `__pclass` value. I found `zend_get_object_classname()` used in core in several places where a class name needed to be reported, so I assume it's fine to use for this exception message.

I can't imagine why folks would want to report a different class name, but I expect the ODM behavior is fine sticking to the `name` value from the struct.